### ISSUE TITLE
Question about the calculation method of loss when there are multiple gpus

### ISSUE THREAD

**Issue by yinaoxiong** [ArrowLuo/CLIP4Clip#101]
https://github.com/ArrowLuo/CLIP4Clip/blob/508ffa3de39ba0563a03199c440ab602a72e9b6f/modules/modeling.py#L400

```
        if self.training:
            visual_output = allgather(visual_output, self.task_config)
            video_mask = allgather(video_mask, self.task_config)
            sequence_output = allgather(sequence_output, self.task_config)
            torch.distributed.barrier()

        visual_output = visual_output / visual_output.norm(dim=-1, keepdim=True)
        visual_output = self._mean_pooling_for_similarity_visual(visual_output, video_mask)
        visual_output = visual_output / visual_output.norm(dim=-1, keepdim=True)

        sequence_output = sequence_output.squeeze(1)
        sequence_output = sequence_output / sequence_output.norm(dim=-1, keepdim=True)

        logit_scale = self.clip.logit_scale.exp()
        retrieve_logits = logit_scale * torch.matmul(sequence_output, visual_output.t())
```
The current code seems to calculate the loss on the global similarity matrix on each gpu. Computing loss only for local and global features as described in https://github.com/openai/CLIP/issues/132 seems to be more computationally and memory efficient.
Sorry to bother you if I misunderstood the code

**Comment by zsnoob** [comment]
My idea is just like yours. After debugging, I found that during the training epoch, all GPUs compute the same global loss with the same sim_matrix instead of individually calculating local losses and then gathering and averaging them. There is a clear computation overlap here. I also have seen that in the function "train_epoch", there is an useless computation "loss.mean()" that seems do nothing after the model.forward(). We only need do local loss following the https://github.com/openai/CLIP/issues/132 and do loss.backward(), The gradient synchronization will be done automatically by DDP.
